In [33]:
import numpy as np
import pandas as pd
import time
from sklearn.datasets import load_svmlight_file

# a9a dataset

## Loading the dataset

### Training set

In [4]:
def get_data(file):
    data = load_svmlight_file(file)
    return data[0], data[1]

X, y = get_data("a9a")
print(X)

  (0, 2)	1.0
  (0, 10)	1.0
  (0, 13)	1.0
  (0, 18)	1.0
  (0, 38)	1.0
  (0, 41)	1.0
  (0, 54)	1.0
  (0, 63)	1.0
  (0, 66)	1.0
  (0, 72)	1.0
  (0, 74)	1.0
  (0, 75)	1.0
  (0, 79)	1.0
  (0, 82)	1.0
  (1, 4)	1.0
  (1, 6)	1.0
  (1, 13)	1.0
  (1, 18)	1.0
  (1, 38)	1.0
  (1, 39)	1.0
  (1, 50)	1.0
  (1, 62)	1.0
  (1, 66)	1.0
  (1, 72)	1.0
  (1, 73)	1.0
  :	:
  (32559, 21)	1.0
  (32559, 35)	1.0
  (32559, 41)	1.0
  (32559, 54)	1.0
  (32559, 61)	1.0
  (32559, 66)	1.0
  (32559, 72)	1.0
  (32559, 73)	1.0
  (32559, 75)	1.0
  (32559, 77)	1.0
  (32559, 82)	1.0
  (32560, 4)	1.0
  (32560, 7)	1.0
  (32560, 17)	1.0
  (32560, 21)	1.0
  (32560, 35)	1.0
  (32560, 39)	1.0
  (32560, 50)	1.0
  (32560, 60)	1.0
  (32560, 66)	1.0
  (32560, 71)	1.0
  (32560, 74)	1.0
  (32560, 75)	1.0
  (32560, 79)	1.0
  (32560, 82)	1.0


In [21]:
# Changing X from sparse representation to standard array
X_train = X.toarray()
y_train = y
print(X_train)
print(y_train)
print(X_train.shape)

[[0. 0. 1. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
[-1. -1. -1. ... -1. -1.  1.]
(32561, 123)


### Test set

In [20]:
X_test, y_test = get_data("a9a.t")
X_test = X_test.toarray()
print(X_test)
print(y_test)
print(X_test.shape)

[[1. 0. 0. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 ...
 [0. 0. 1. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]]
[-1. -1.  1. ... -1. -1.  1.]
(16281, 122)


## Preprocessing

In [23]:
from sklearn.decomposition import PCA

pca = PCA(n_components=14)

### Traning data

In [24]:
X_train_reduced = pca.fit_transform(X_train)
X_train_reduced

array([[ 0.23239737, -0.68819295,  0.85667136, ...,  0.02148147,
         0.15861185,  0.389433  ],
       [-1.20662065, -0.85959757,  0.616277  , ...,  0.06610164,
         0.4327976 , -0.28969152],
       [ 0.2116386 ,  1.22573661,  0.17644706, ...,  0.69390972,
        -0.56824374,  0.41972188],
       ...,
       [ 0.74332647,  1.26675186,  0.60426759, ..., -0.11812197,
        -0.10765446,  0.24485983],
       [ 0.87047487,  0.93608506, -0.76327228, ...,  0.74868369,
        -0.29233675,  0.04465699],
       [-0.12202918,  0.93561235,  0.72740491, ..., -0.39892826,
        -0.18836753, -0.8295538 ]])

### Testing set

In [25]:
X_test_reduced = pca.fit_transform(X_test)
X_test_reduced

array([[ 0.81663618,  0.02498892, -0.87664761, ..., -0.34917539,
         0.57748621, -0.03951886],
       [-1.16837675,  0.91244668, -0.10008568, ...,  0.05053122,
         0.1504557 , -0.2789483 ],
       [-0.9721929 , -0.19001809, -0.19611116, ...,  0.06833835,
        -0.32440394,  0.03113935],
       ...,
       [-1.26055427, -0.86482666,  0.62352845, ..., -0.1280389 ,
        -0.16634468, -0.14126516],
       [ 0.06726798, -0.51248111,  0.70476954, ...,  0.6600293 ,
        -0.46138724, -0.14651061],
       [-1.38988745, -0.90240615,  0.70796896, ..., -0.64109149,
         0.0622353 ,  0.65524468]])

## Applying passive agressive online learning

In [16]:
from sklearn.metrics import accuracy_score
from passive_agressive_classic_update import passive_agressive_online

### Classic update

In [26]:
w_classic = passive_agressive_online(X_train_reduced, y_train, impl="classic")
w_classic

array([-0.81809909, -0.23873789,  0.32811158, -0.15514349, -0.39851645,
       -0.18825116, -0.82564869, -1.4156379 , -0.24369112, -0.84731157,
       -0.40837704, -1.04690991, -1.05513819,  0.12503183])

In [27]:
# Accuracy on the training set
predicted = np.sign(np.dot(X_train_reduced, w_classic))
accuracy = accuracy_score(predicted, y_train)
print("%.2f" % (accuracy*100))

64.81


Classic update is barely better than random guessing (64.81%). It's not an outstanding performance. Let's have a look at the test performance

In [29]:
# Accuracy on the test set
predicted = np.sign(np.dot(X_test_reduced, w_classic))
accuracy = accuracy_score(predicted, y_test)
print("%.2f" % (accuracy*100))

65.23


Classic update has a performance close to its training data performanced on unseen data as well (65.23%). Might it be underfitting?
Let's have a look at first relaxation

### First relaxation

In [32]:
# Generating C values
C = []
for i in range(-15, 1):
    C.append(10**i)
C

[1e-15,
 1e-14,
 1e-13,
 1e-12,
 1e-11,
 1e-10,
 1e-09,
 1e-08,
 1e-07,
 1e-06,
 1e-05,
 0.0001,
 0.001,
 0.01,
 0.1,
 1]

In [39]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold

kf = KFold(n_splits=5)
C_scores = np.zeros((len(C), 5))
C_times = np.zeros((len(C), 5))
current_fold = 0
for train, test in kf.split(X_train_reduced):
    X_train, X_test, y_train_kf, y_test_kf = X_train_reduced[train], X_train_reduced[test], y_train[train], y_train[test]
    for i in range(len(C)):
        begin = time.time()
        w_relax1 = passive_agressive_online(X_train, y_train_kf, impl="relax1", C=C[i])
        end = time.time()
        predicted = np.sign(np.dot(X_test, w_relax1))
        accuracy = accuracy_score(predicted, y_test_kf)
        C_scores[i, current_fold] = accuracy
        C_times[i, current_fold] = end - begin
    current_fold += 1
print(C_scores)
print(C_times)

[[0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.7076616  0.70684889 0.71114865 0.71038084 0.7105344 ]
 [0.71150008 0.71022727 0.71283784 0.71268428 0.71268428]
 [0.71027176 0.70669533 0.70823096 0.71022727 0.70715602]
 [0.71242131 0.70884521 0.71022727 0.71114865 0.70408477]
 [0.69491786 0.68596437 0.69195332 0.69963145 0.68212531]
 [0.65392292 0.64404177 0.63835995 0.65433047 0.64189189]]
[[1.82618523 1.80763674 1.78003359 1.91355038 1.8363235 ]
 [2.12645078 

In [53]:
# Selecting the C with the best average accuracy
C_scores_avg = np.mean(C_scores, axis=1)
C_times_avg = np.mean(C_times, axis=1)
print(C_scores_avg)
best_index = np.argmax(C_scores_avg)
print("The best value for the parameter C is:", C[best_index])
print("Its average accuracy is %.2f" % (C_scores_avg[best_index]*100))
print("The average running time of passive-agressive approach first relaxation with parameter C =", C[best_index],
     "is %.2f s" % C_times_avg[best_index])

[0.70931488 0.70931488 0.70931488 0.70931488 0.70931488 0.70931488
 0.70931488 0.70931488 0.70931488 0.70931488 0.70931488 0.71198675
 0.70851627 0.70934544 0.69091846 0.6465094 ]
The best value for the parameter C is: 0.0001
Its average accuracy is 71.20
The average running time of passive-agressive approach first relaxation with parameter C = 0.0001 is 1.91 s


The first relaxation with C=0.0001 has a good accuracy on the cross-validation set (averaging 71.2%) but at the expense of a higher running time (1.91s on average).

In [ ]:
# Use it on test set
predicted = np.sign(np.)